File: algorithms/2_medium/R/00648_replace_words/solution.cpp
class Solution {
public:
    static string replaceWords(vector<string>& dictionary, const string& sentence) {
        ranges::sort(dictionary, [](const auto& a, const auto& b) {
            return a.size() < b.size();
        });

        ostringstream res{};
        istringstream ss{sentence};
        string word;

        while (ss >> word) {
            bool is_inserted {false};

            for (const auto& d : dictionary) {
                if (word.starts_with(d)) {
                    res << d << ' ';
                    is_inserted = true;
                    break;
                }
            }

            if (!is_inserted) {
                res << move(word) << ' ';
            }
        }

        string final_word {res.str()};

        if (!final_word.empty()) {
            final_word.pop_back(); // to remove last ' ' char
        }

        return final_word;
    }
};

File: algorithms/2_medium/R/00648_replace_words/solution.py
class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        dictionary.sort(key=lambda x: len(x))
        res = []

        for word in sentence.split():
            is_inserted = False

            for d in dictionary:
                if word.startswith(d):
                    res.append(d)
                    is_inserted = True
                    break

            if not is_inserted:
                res.append(word)

        return " ".join(res)
